### ISSUE TITLE
Repeated WAL entries when WriteOptions.sync=true is specified

### ISSUE THREAD

**Issue by itsbilal** [facebook/rocksdb#7576]
At CockroachDB, we have built a sync loop that calls `FlushWAL(sync=true)`  on regular intervals:

https://github.com/cockroachdb/cockroach/blob/c0f7137093a5a08825ca1614841439a812ae7216/pkg/storage/rocksdb.go#L360

We recently made a change to pass in `WriteOptions.sync = true` to all non-batch mutation operators, and we started seeing sporadic failures in many of our integration tests, with some having the error `Corruption: Sequence numbers in WAL not in increasing order: 8759 < 8767`

Taking a deeper dive, we noticed that some WALs had repeated records in the middle: 

```
19901447(50) seq=3790026 count=1
19901508(138) seq=3790027 count=2
19901657(840) seq=3790029 count=1
19902508(51) seq=3790030 count=1
19902570(50) seq=3790026 count=1
19902631(138) seq=3790027 count=2
19902780(840) seq=3790029 count=1
19903631(51) seq=3790030 count=1
```

(Every line above is a WAL record, corresponding to one batch)

### Expected behavior

WAL shouldn't have repeated records, and RocksDB should be correctly handling synchronization in this use case, or alternatively, erroring out if an unsupported concurrent request is made.

### Actual behavior

The WAL contains repeated records, that cause errors upon WAL replay

**Comment by pdillinger** [comment]
Looks like we need to add FlushWAL to our stess test

**Comment by ajkr** [comment]
Thanks for reporting! Should be finally fixed by #8746
